README
= GPMD85Emulator ==============================================================
- multiplatform GNU/GPL emulator of the Tesla PMD 85, an 8-bit personal micro-
- computer produced in eighties of twentieth century in former Czechoslovakia

AUTHORS:
~ mborik [Martin Borik] - leader programmer of this *nix port based on SDL
~ rombor [Roman Borik]  - original Windows code and emulation of CPU and chips
~ ub880d                - support developer, lead-tester and bug-finder

THANKS:
~ ikon/SinDiKat   - translation of comments in classes from slovak language
~ mikezt/zeroteam - quick-search in file-selector and beta-testing
~ Staon           - fixing compatibility bug in file-selector's ScanDir

REQUIRED LIBRARIES:
~ SDL - Simple DirectMedia Library  [ libSDL >= 1.2 ]

HOTKEYS:
~ function [f] keys are any of Alt|Win|Mac|Meta keys
~ main menu appear with [f]+F1 or with the "menu key"
~ for start/stop of tape emulator use [f]+P hotkey

INSTALLATION:
~ run script './configure', then 'make'
~ you can install application to the system dirs with 'make install' under root
~ in case of auto-config problems, you can try to 'autoreconf -vf',
  or 'automake -af' (but you need autotools package).

CONFIGURATION PARAMETERS:
~ to enable/disable OpenGL functionality you need to run configuration with
  optional parameter './configure --with-opengl' or '--without-opengl'.
~ to enable trace informations use '--enable-trace' parameter in configure.



ChangeLog
2018-03-01 [mborik] writing into VRAM memspace is projecting into union rectangle and screen is rendered only if rect isn't empty
2018-03-01 [mborik] bugfix of regression introduced in last commit when ROM modules didn't work (found by Cizo, thanks!)
2018-02-13 [mborik] fixed long-time persisting issue with changing memory/rom/rmm configuration; fixed bug in ChipMemory2AEx
2018-02-12 [mborik] implemented 256k memory expansion; fixed updir path related bug; striping of long filenames in menu
2018-02-05 [mborik] settings updated and extended to all new features; slim down of roms - removed obsolete and unusable
2018-01-30 [mborik] ChipMemory splitted into separate child classes for each model, method for reset status & currently mapped VRAM pointer
2018-01-22 [mborik] implemented ChipCpu8080P which take into account slowdown of the video-processor
2018-01-21 [mborik] fixed illegal memory access bug in x4 scale function
2018-01-20 [mborik] fixed critical issues with window/renderer/screen (re)initialization
2018-01-14 [mborik] scalers replaced with scanliner texture overlay
2018-01-13 [mborik] fixed all menu and keyhandling issues after SDL2 migration
2018-01-11 [mborik] drop of OpenGL hacks, fixed init crack in sound driver, implemented surface blitter to fix icons in statusbar
2018-01-10 [mborik] massive refactoring of screen and user interface processing and rendering
2018-01-08 [mborik] upgraded SDL to 2.x, simplified screen rendering (unstable, missing menu overlay & scanliners)
2017-12-21 [mborik] refactoring of ScreenPMD85::FillBuffer and keyword multicolor to proper ColorAce
2017-12-20 [mikezt] fixed disk image saving problems
2017-09-18 [mborik] warnings suppression (phase 2), project update, repo cleanup and fixing the build for MacOS
2017-09-14 [mborik] warnings suppression (phase 1)
2017-04-19 [ub880d] bugfix in screen rendering and fixing the build for new gcc (narrowing conversions)
2016-07-25 [ub880d] fixed missing callback on menu entry "OK" of "SAVE MEMORY"
2015-05-24 [ub880d] added warning about wrong file extension for ROM modules
2015-04-19 [ub880d] fixed labels/shortcuts in menu 'FILE'
2014-02-24 [Staon]  project tracker ticket #2: ScanDir compatibility bug
2012-06-19 [mborik] fixed crashes during OpenGL initialization with reinterpret_cast
2012-06-18 [mborik] prerelease bugfixes and mods
2012-06-08 [mborik] fixed multithreaded video rendering problems (especially in OpenGL) with SDL_VIDEOEXPOSE event
2012-06-07 [mborik] implemented read/write memory block; fixed some time-critical bugs in dialogues
2012-06-05 [mborik] imported rombor's improvements in emulation of 8253, ifTimer and PMD32; removed ifMusica occurrences
2012-06-04 [mborik] implemented emulation speed change
2012-06-01 [ikon]   comment translation sk>en (1st phase)
2012-05-31 [mborik] fullscreen toggle; some minor fixes
2012-05-30 [mborik] reworked status bar icons loading and some palette problems solved
2012-05-26 [mborik] bugfix and fine-tuning of OpenGL screen rendering
2012-05-24 [mborik] OpenGL screen rendering implementation which can be enabled with configure (--with-opengl)
2012-03-15 [mborik] instruction tracing in debugger (too buggy yet)
2012-03-04 [mborik] first implementation of debugger (drawings only)
2012-02-22 [mborik] improvements and some mods on quick-searching in fileSelector (with some ideas by ub880d :)
2012-02-21 [mikezt] implemented quick-searching in fileSelector
2012-02-18 [mborik] major bugfixes in Settings class and finalized saving of settings (autosave too :)
2012-02-11 [mborik] fixed TapeBrowser bug if tape file from configuration not exists
2012-02-05 [mborik] implemented file import to TapeBrowser and extended dialog with hex/dec shortcut
2012-02-04 [mborik] RGBA mask workaround added to screen surface creation for big endian compatibility
2012-02-02 [mborik] separation of GUI vs. Core code; GUI code separated to logical sections; implemented move of tape block or selection
2012-02-01 [mborik] implemented "pause on lost focus"; added "autosave settings" to menu; fixed callbacks of TapeBrowser popup items
2012-01-31 [mborik] externed GUI, Settings and TapeBrowser to make it global accessible
2012-01-27 [mborik] implemented methods for checking selection continuity and deleting selection to TapeBrowser
2012-01-23 [mborik] completely Valgrinated project, solved all memleaks and other bugs; implemented tape block selection with insert key
2012-01-22 [mborik] configuration file changed to plain text .conf file, libXML2 not yet required!
2012-01-06 [mborik] early implementation of context popup menu for block operations in TapeBrowser (not functional yet :)
2012-01-05 [mborik] other fixes and optimalizations in TapeBrowser
2012-01-04 [mborik] fully functional saving to tape, new menu item "New tape", prepared dialogues "Save changes?" for unsaved changes
2012-01-03 [mborik] fixed segfaults while freeing fileSelector memcache and disabled recacheing while changing checkboxes in fileSelector and TapeBrowser
2012-01-02 [mborik] finished prenatal version of TapeBrowser, added tape activity progress to statusbar
2012-01-01 [mborik] fully functional tape emulation, added current model acronym and tape activity icon to status bar
2011-12-30 [mborik] fixed extended 384px width screen mode for Consul 2717 (and clean-up of trivial inline functions in SystemPIO)
2011-12-28 [ub880d] show-hidden-files in filebrowser (hotkey switch META-PERIOD), prepared configuration file for auto-save-config, configuration file version changed to 1.1
2011-12-18 [ub880d] 4x4 screen size and scalers
2011-12-18 [mborik] tracing statements in video/sound initialization and can be enabled with configure (--enable-trace)
2011-12-17 [mborik] first public alpha version


src/Pmd32.h
/*	Pmd32.h: Class for emulation of disk drive PMD 32
	Copyright (c) 2008-2010 Roman Borik <pmd85emu@gmail.com>
	Copyright (c) 2011 Martin Borik <mborik@users.sourceforge.net>

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program. If not, see <http://www.gnu.org/licenses/>.
*/
//---------------------------------------------------------------------------
#ifndef Pmd32H
#define Pmd32H
//---------------------------------------------------------------------------
#include "CommonUtils.h"
#include "ChipMemory.h"
#include "IifGPIO.h"
//---------------------------------------------------------------------------
#define OPER_FREQ               20000
#define CLOCK_PERIOD            (CPU_FREQ / OPER_FREQ)
#define PRESENT_TICKS           100     // 5 ms
#define SEND_BYTE_TICKS         2       // 100 us
#define FORMAT_RES_TICKS        8000    // 400 ms
#define RDWR_RES_TICKS          20      // 1 ms

#define PRESENTATION            0xAA
#define ACK                     0x33
#define NAK                     0x99

#define RESULT_OK               0       // OK
#define RESULT_WP               1       // write-protected disk
#define RESULT_FE               2       // format error
#define RESULT_RE               3       // read error
#define RESULT_WE               4       // write error
#define RESULT_BD               5       // bad disk number
#define RESULT_NF               6       // file/dir not found
#define RESULT_TL               7       // file/dir name too long
#define RESULT_CE               8       // disk image creation error
#define RESULT_CM               9       // configuration file missed
#define RESULT_CR               10      // configuration file read error
#define RESULT_CC               11      // configuration file corrupted
#define RESULT_UI               12      // unknown disk image file

#define IDLE_STATE              0
#define WAIT_PRESENT            1
#define WAIT_COMMAND            2
#define WAIT_CRC                3
#define WAIT_SECTOR             4
#define WAIT_TRACK              5
#define WAIT_DRIVE              6
#define WAIT_DATA               7
#define SEND_DATA               8
#define SEND_CRC                9
#define SEND_ACK                10
#define SEND_RESULT             11
#define SEND_NAK                12
#define WAIT_ADDR_H             13
#define WAIT_ADDR_L             14
#define WAIT_LEN_H              15
#define WAIT_LEN_L              16
#define WAIT_DATA_MEM           17
#define SEND_DATA_MEM           18
#define WAIT_FIND_TYPE          19
#define WAIT_LEN_NAME           20
#define WAIT_DATA_NAME          21
#define WAIT_WP                 22

#define SECTOR_SIZE             128
#define PHYSICAL_SECTOR_SIZE    (4 * SECTOR_SIZE)
#define INTERNAL_RAM_SIZE       (8 * SECTOR_SIZE)

#define MAX_SECTORS_PER_TRACKS  64
#define MAX_TRACKS              256

#define NUM_DRIVES              4
#define DRIVE_A                 0
#define DRIVE_B                 1
#define DRIVE_C                 2
#define DRIVE_D                 3

#define SDROOT_IMAGES_DEF_FILE  "images.cfg"
//---------------------------------------------------------------------------
class Pmd32 : public sigslot::has_slots<>
{
	public:
		Pmd32(IifGPIO *pio);
		virtual ~Pmd32();

		void SetExtraCommands(bool extraCommands, char *sdPath);
		void BeforeReset();
		void AfterReset();
		void OnSetMode2(BYTE CWR);
		void OnHandshake();

		void Disk32Service(int ticks, int dur);
		int  InsertDisk(int drive, char *file, bool WP);
		void RemoveDisk(int drive);
		inline void SetWpDisk(int drv, bool wp) { drives[drv].wp = wp; }

		int diskIcon;

	private:
		typedef struct {
			char *filePath;
			FILE *handle;
			bool  wp;
			BYTE  tracks;     // total tracks count
			BYTE  sectors;    // logical sectors per track
			BYTE  sectorSize; // physical sector size (0=128B, 1=256B, 2=512B, 3=1KB)
		} DRIVES;

		IifGPIO *pio;
		DRIVES drives[NUM_DRIVES];

		BYTE  drvnum;
		BYTE  track;
		BYTE  sector;
		int   address;
		int   length;
		BYTE  findType;
		BYTE  wp;

		BYTE  memory[INTERNAL_RAM_SIZE];
		BYTE  buffer[MAX_SECTORS_PER_TRACKS * SECTOR_SIZE];
		BYTE *point;
		DWORD dwRW;

		BYTE  command;
		BYTE  CRC;
		int   byteCounter;
		bool  result;
		bool  inHandshake;

		BYTE  toSend;
		bool  noSend;

		int   diskState;
		int   diskTicks;
		int   tickCounter;

		bool  extraCommands;
		char *sdRoot;
		char *sdPath;
		char *sdFile;
		DIR  *hFind;
		char *hFindDir;

		void  Disk32ServiceStateCheck();
		void  Disk32ServiceSendResultCommand();

		bool  FindFile();
		bool  PrepareSector();
		bool  WriteSector();
		int   CheckImageType(char *fn, BYTE *trk, BYTE *sec, BYTE *phys);
		bool  LocateFileInPathCaseInsensitive(char *filePath);
};
//---------------------------------------------------------------------------
#endif
//---------------------------------------------------------------------------
